kubeflow/core/jupyterhub_config.py

```diff
@@ -0,0 +1,141 @@
+import os
+from importlib.util import spec_from_file_location, module_from_spec
+from jhub_remote_user_authenticator.remote_user_auth import \
+    RemoteUserAuthenticator
+
+SERVICE_ACCOUNT_SECRET_MOUNT = '/var/run/secrets/sa'
+
+spec = spec_from_file_location('spawner', '/etc/config/spawner.py')
+spawner = module_from_spec(spec)
+spec.loader.exec_module(spawner)
+
+###################################################
+# JupyterHub Options
+###################################################
+c.JupyterHub.ip = '0.0.0.0'
+c.JupyterHub.hub_ip = '0.0.0.0'
+# Don't try to cleanup servers on exit - since in general for k8s, we want
+# the hub to be able to restart without losing user containers
+c.JupyterHub.cleanup_servers = False
+###################################################
+
+###################################################
+# Spawner Options
+###################################################
+c.JupyterHub.spawner_class = spawner.KubeFormSpawner
+
+c.KubeSpawner.cmd = 'start-singleuser.sh'
+c.KubeSpawner.args = ['--allow-root']
+# gpu images are very large ~15GB. need a large timeout.
+c.KubeSpawner.start_timeout = 60 * 30
+# Increase timeout to 5 minutes to avoid HTTP 500 errors on JupyterHub
+c.KubeSpawner.http_timeout = 60 * 5
+
+# Volume setup
+c.KubeSpawner.singleuser_uid = 1000
+c.KubeSpawner.singleuser_fs_gid = 100
+c.KubeSpawner.singleuser_working_dir = '/home/jovyan'
+volumes = []
+volume_mounts = []
+
+# Allow environment vars to override uid and gid.
+# This allows local host path mounts to be read/writable
+env_uid = os.environ.get('NOTEBOOK_UID')
+if env_uid:
+    c.KubeSpawner.singleuser_uid = int(env_uid)
+env_gid = os.environ.get('NOTEBOOK_GID')
+if env_gid:
+    c.KubeSpawner.singleuser_fs_gid = int(env_gid)
+access_local_fs = os.environ.get('ACCESS_LOCAL_FS')
+if access_local_fs == 'true':
+    def modify_pod_hook(spawner, pod):
```

Comment:
Is it possible to add set more pod attributes directly the way we're doing here? For example volumes, volume_mounts.

Reply:
Kam, thanks for taking the time to review this!

I'm not sure I understand the workflow you have in mind here.
The original `kubeform_spawner.py` file had code which we have left intact, below, to compute values for `volumes`, `volume_mounts`, and this PR does not interfere with this piece of pre-existing functionality.

So, can you give an example of what you have in mind and how the administrator would be able to use it?

One way we have been thinking about allowing the administrator to set generic notebook pod attributes is:
1. the administrator could be able to modify an initial pod definition, e.g., in a file named `<uidir>/pod.yaml`, so they can specify any attribute they want in a generic way
2. we could expose this file inside the JupyterHub
3. the UI could start from this definition, and apply to it all decisions made by the user via the form before spawning the pod.

But this seems to require changes all the way to `make_pod()`  in KubeSpawner, see https://github.com/jupyterhub/kubespawner/blob/f49b1bd7060d0ca6fdf3342bb47491f33f95d79d/kubespawner/objects.py#L220, so we feel it was a much more disruptive change, and definitely outside the scope of an initial PR. We can always tackle this as a next step of course.